rfcs/0030-review-workflow.md

```diff
@@ -0,0 +1,132 @@
+`needs:work` means that the PR is not ready to be merged as-is and needs to be
+improved. `needs:review` means that all the work is done and somebody (usually
+the PR creator) thinks that it should be ready for merge. Reviewers (with or
+without commit access) can then go through the list of `needs:review` PRs and
+review them. They can than either request changes and set the state to
+`needs:work` or merge the PR. In case of review by non-commiters, we could
+introduce an additional state like `needs:merge`. Reviewing and merging those
+PRs should then take little time for the committers and could double as a sort
+of training program for potential new committers.
+
+This workflow is not currently possible due to GitHub's restrictions. Namely
+only people with commit access can set tags on a PR. This can be resolved
+through tooling. Work on this is [in progress](https://github.com/NixOS/ofborg/pull/216).
+
+All the above issues should be resolved by this:
+
+1. There is no reason anymore to preferably review recent PRs. Before, recent
+   commit were the only ones where one can be sure that they are not reviewed
+   yet (and thus in a `needs:work` state). Now, reviewers can randomly choose
+   and `needs:review` PR.
+
+2. Similarly, the reviewer may change at any time. If the first reviewer forgets
+   or loses interest in a PR, the PR will just end up in the `needs:review` list
+   again and can easily find a new reviewer.
+
+3. People without commit access can help reviewing commits and lessen the load
+   on committers without putting the reviewed PRs at a disadvantage.
+
+4. We can easily gather statistics. In the long run, we might even implement
+   some sort of automated triage where we regularly ping `needs:work` PRs and
+   close them after the authors went unresponsive for a while. Or we could
+   tag them as `needs:takeover`. But this is out of scope.
+
+
+An example PR might go like this. `contributor` creates the PR, `reviewer1`
+reviews without commit access, `reviewer2` reviews with commit access.
+
+> contributor: hello: 42.0 -> 43.0 [PR text]
+
+[the bot automatically sets the new PR to `needs:review`]
+
+> reviewer1: Looks good to me. Just one little nitpick: Please add a waving emoji to the name.
+> @GrahamcOfBorg status needs:work
+
+> contributor: Good point, more emojis are always good. Done.
+> @GrahamcOfBorg status needs:review
+
+> reviewer1: Thanks! I think this is ready for merge.
+> @GrahamcOfBorg status needs:merge
+
+> reviewer2: LGTM
+> @GrahamcOfBorg build-and-merge
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Contributing would be a bit more complicated. There is one more bit of
+  knowledge new contributors need. It is possible that some contributors forget
+  to set the PR to `needs:review` again after a fixup. That might be resolved
+  by some form of automatic triage as explained above. It would also be not much
+  than the current status, since the first reviewer will still be notified by
+  GitHub.
```

Comment:
Is it possible to figure out if the `First contributor` batch on github is set for a pull request? Then additional help could be shown to first timers.

Reply:
Yes, that would be cool. @grahamc do you know if the GitHub api allows that?

Reply:
I think it has to be a setting, not even API — this should change what is shown _before_ submission of the PR, there is no event to be notified about at that point…

After submission The Tool could check if the submitter is known and post a link/summary of review process.